File: Case-specific Sorting of Strings - GFG/casespecific-sorting-of-strings.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution
{
    public:
    //Function to perform case-specific sorting of strings.
    string caseSort(string s, int n)
    {
        // your code here
        vector<int>dp(n,0);
        vector<char>cap,sml;
        for(int i=0;i<n;i++){
            if(s[i]>='a' && s[i]<='z'){
                sml.push_back(s[i]);
            }
            else{
                cap.push_back(s[i]);
                dp[i]=1;
            }
        }
        sort(cap.begin(),cap.end());
        sort(sml.begin(),sml.end());
        int i=0,j=0,k=0;
        while(i<n){
            if(dp[i]){
                s[i]=cap[j];
                j++;
            }
            else{
                s[i]=sml[k];
                k++;
            }
            i++;
        }
        return s;
    }
};

//{ Driver Code Starts.

int main()
{
	int t;
	cin>>t;
	while(t--)
	{
		int n;
		cin>>n;
		string str;
		cin>>str;
		Solution obj;
		cout<<obj.caseSort (str, n)<<endl;
	}
}
// } Driver Code Ends

File: Case-specific Sorting of Strings - GFG/casespecific-sorting-of-strings.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;


// } Driver Code Ends
//User function Template for Java

class Solution
{
    //Function to perform case-specific sorting of strings.
    public static String caseSort(String str)
    {
        char[] st = new char[str.length()];
        int k=0,j=0;
        char a[]=str.toCharArray();
        Arrays.sort(a);
        while(j<a.length){
            if(a[j]>='a')
                break;
            j++;
        }
        for(int i=0;i<str.length();i++){
            if(Character.isLowerCase(str.charAt(i))){
                st[i]=a[j];
                j++;
            }
            else{
                st[i]=a[k];
                k++;
            }
        }
        return new String(st);
    }
}


//{ Driver Code Starts.

class GFG {
	public static void main (String[] args) {
		
    	try {
    	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	    PrintWriter out=new PrintWriter(System.out);
    	    String[] words = br.readLine().split("\\s+");
    	    int numTestCases = Integer.parseInt(words[0]);
    	    
    	    for (int tIdx = 0; tIdx < numTestCases; tIdx++) {
    	        words = br.readLine().split("\\s+");
    	        int size = Integer.parseInt(words[0]);
    	        String str = br.readLine();
    	        String sortedStr = new Solution().caseSort(str);
    	        out.println(sortedStr);
    	    }
    	    out.close();
    	}
    	catch (IOException e) {
    	    System.out.println(e);
    	}
	}
}
// } Driver Code Ends